compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,436 @@
+          let curr = propertyNode.parent;
+          while (curr != null) {
+            assumedNonNullObjects.add(curr);
+            curr = curr.parent;
+          }
+        }
+      } else if (value.kind === "LoadLocal") {
+        if (
+          lvalue.identifier.name == null &&
+          value.place.identifier.name !== null &&
+          !usedOutside
+        ) {
+          c.declareTemporary(lvalue.identifier, value.place.identifier);
+        }
+      }
+      /**
+       * Note that we do not record StoreLocals as this runs after ExitSSA.
+       * As a result, an expression like `(a ?? b).c` is represented as two
+       * StoreLocals to the same identifier id.
+       */
+    }
+
+    if (
+      block.terminal.kind === "scope" ||
+      block.terminal.kind === "pruned-scope"
+    ) {
+      scopeStartBlocks.set(block.terminal.block, block.terminal.scope.id);
+    }
+
+    nodes.set(blockId, {
+      blockId,
+      scope: scopeStartBlocks.get(blockId) ?? null,
+      preds: block.preds,
+      assumedNonNullObjects,
+    });
+  }
+  return nodes;
+}
+
+function deriveNonNull(fn: HIRFunction, nodes: Map<BlockId, BlockInfo>): void {
+  // block -> successors sidemap
+  const succ = new Map<BlockId, Set<BlockId>>();
+  const terminalPreds = new Set<BlockId>();
+
+  for (const [blockId, block] of fn.body.blocks) {
+    for (const pred of block.preds) {
+      const predVal = getOrInsertDefault(succ, pred, new Set());
+      predVal.add(blockId);
+    }
+    if (block.terminal.kind === "throw" || block.terminal.kind === "return") {
+      terminalPreds.add(blockId);
+    }
+  }
+
+  function recursivelyDeriveNonNull(
+    nodeId: BlockId,
+    kind: "succ" | "pred",
+    traversalState: Map<BlockId, "active" | "done">,
+    result: Map<BlockId, Set<PropertyLoadNode>>
+  ): boolean {
+    if (traversalState.has(nodeId)) {
+      return false;
+    }
+    traversalState.set(nodeId, "active");
+
+    const node = nodes.get(nodeId);
+    if (node == null) {
+      CompilerError.invariant(false, {
+        reason: `Bad node ${nodeId}, kind: ${kind}`,
+        loc: GeneratedSource,
+      });
+    }
+    const neighbors = Array.from(
+      kind === "succ" ? succ.get(nodeId) ?? [] : node.preds
+    );
+
+    let changed = false;
+    for (const pred of neighbors) {
+      if (!traversalState.has(pred)) {
+        const neighborChanged = recursivelyDeriveNonNull(
+          pred,
+          kind,
+          traversalState,
+          result
+        );
+        changed ||= neighborChanged;
+      }
+    }
+    // active neighbors can be filtered out as we're solving for the following
+    // relation.
+    // X = Intersect(X_neighbors, X)
+    // non-active neighbors with no recorded results can occur due to backedges.
+    // it's not safe to assume they can be filtered out (e.g. not intersected)
+    const neighborAccesses = Set_intersect([
+      ...(Array.from(neighbors)
+        .filter((n) => traversalState.get(n) !== "active")
```

Comment:
nit: it's easier to read a positive comparison, ie `traversalState.get(n) === 'done'`